elytron/ELY-2078-encryption-for-FilesystemSecurityRealm.adoc

```diff
@@ -0,0 +1,222 @@
+
+For the default out of box configuration for WildFly, the goal is for the FileSystemSecurityRealm to make use of ScramDigest with AES encryption and integriy enabled.
+
+== Issue Metadata
+
+=== Issue
+
+* https://issues.redhat.com/browse/ELY-2078[ELY-2078]
+* https://issues.redhat.com/browse/WFCORE-5530[WFCORE-5530]
+* https://issues.redhat.com/browse/WFLY-15075[WFLY-15075]
+
+
+=== Related Issues
+
+* https://issues.redhat.com/browse/EAP7-1717[EAP7-1717]
+
+=== Dev Contacts
+
+* mailto:{email}[{author}]
+
+=== QE Contacts
+
+=== Testing By
+* [X] Engineering
+
+* [ ] QE
+
+=== Affected Projects or Components
+- Widlfly, Wildfly Elytron, Wildfly Core
+
+=== Other Interested Projects
+N/A
+
+== Requirements
+=== Filesystem Encryption
+==== Hard Requirements
+In the FileSystemSecurityRealm in Elytron, a new optional parameter is introduced to the constructor that accepts a SecretKey. Passing in that parameter indicates that the user wants to encrypt the FileSystem. Also to eliminate the number of overloaded constructors, a FileSystemSecurityRealmBuilder was introduced.
+This is an example of using the new FileSystemSecurityRealm builder.
+```java
+ FileSystemSecurityRealm.builder()
+                        .setRoot(getRootPath())
+                        .setLevels(3)
+                        .setSecretKey(key)
+                        .build();
+```
+This will create a new file system with 3 levels for the identity file, encrypted with the Secret Key "key".
+
+In WildFly Core, the filesystem-realm will include new attributes to specify the credential store and the alias within the store that holds the secret key that should be used to enable encryption for the filesystem realm.
+
+*Attributes*
+
+`credential-store` - a reference to the credential store that contains the secret key used to encrypt and decrypt the filesystem-realm. This is an optional attribute.
+
+`secret-key` - an alias to the secret key used to encrypt and decrypt the filesystem-realm. If the credential-store attribute is specified, this attribute is required. Otherwise this is an optional attribute.
+
+To create an encrypted FileSystem the following commands would be used
+
+```bash
+/subsystem=elytron/secret-key-credential-store=mycredstore:add(path=propcredstore.cs, relative-to=jboss.server.config.dir, create=true, populate=true)
+```
+```bash
+/subsystem=elytron/filesystem-realm=fsRealm:add(path=fs-realm-users, relative-to=jboss.server.config.dir, credential-store=propcredstore.cs, secret-key=key)
+```
+The filename is stored as the the principal (also know as the username) encoded in base32, once an identity is created with encryption enabled.
```

Comment:
Does this mean that the `encoded` attribute has to be `true` when the `credential-store` and `secret-key` are set?

Reply:
In my implementation, if `credential-store` and `secret-key` are set, then the `encoded` gets set to false and the whole prinicpal is encoded with base32 (eg: the principal plainUser is changed to OBWGC2LOKVZWK4Q).

Whereas if `encoded` is set to true, the prinicpal is `{USER}-{USER_BASE32}` (eg: the principal plainUser is changed to `plainUser-OBWGC2LOKVZWK4Q`)